lmdeploy/serve/qos_engine/qos_engine.py

```diff
@@ -0,0 +1,241 @@
+                    request.session_id + i,
+                    True,  # always use stream to enable batching
+                    sequence_start=True,
+                    sequence_end=True,
+                    request_output_len=request.max_tokens
+                    if request.max_tokens else 512,
+                    stop=False,
+                    top_p=request.top_p,
+                    temperature=request.temperature,
+                    repetition_penalty=request.repetition_penalty,
+                    ignore_eos=request.ignore_eos,
+                    do_preprocess=False)
+                generators.append(result_generator)
+            return generators
+
+        elif isinstance(request, GenerateRequestQos):
+            async_engine = self.engine
+            sequence_start = async_engine.steps.get(str(request.session_id),
+                                                    0) == 0
+            sequence_end = not request.interactive_mode
+
+            generation = async_engine.generate(
+                request.prompt,
+                request.session_id,
+                stream_response=True,  # always use stream to enable batching
+                sequence_start=sequence_start,
+                sequence_end=sequence_end,
+                request_output_len=request.request_output_len,
+                top_p=request.top_p,
+                top_k=request.top_k,
+                stop=request.stop,
+                temperature=request.temperature,
+                repetition_penalty=request.repetition_penalty,
+                ignore_eos=request.ignore_eos)
+            return generation
+
+        elif isinstance(request, ChatCompletionRequestQos):
+            # default chat/completions
+            result_generator = self.engine.generate(
+                request.messages,
+                request.session_id,
+                True,  # always use stream to enable batching
+                sequence_start=True,
+                sequence_end=True,
+                request_output_len=request.max_tokens
+                if request.max_tokens else 512,
+                stop=request.stop,
+                top_p=request.top_p,
+                temperature=request.temperature,
+                repetition_penalty=request.repetition_penalty,
+                ignore_eos=request.ignore_eos)
+            return result_generator
+
+        return time.sleep(0.01)
+
+    async def generate_with_qos(self, request):
+        if not self.is_qos_enabled():
+            return await self.generate(request)
+
+        # push (request,event) to queue
+        event = asyncio.Event()
+        request_event = (request, event)
+        with self.lock:
+            self.qos_user_group.enqueue(request_event)
+
+        await event.wait()
+
+        result_generator = await self.generate(request)
+
+        # release self.availSlots resources
+        with self.lock:
+            if isinstance(request, CompletionRequestQos):
```

Comment:
Is `request.prompt` a list or a string?

Reply:
if request is  `CompletionRequestQos`, request.prompt is a list ,else a string

Reply:
CompletionRequestQos accepts `str` and `List[str]` both.

Reply:
We will check if `request.prompt` is `str` or `List[str]`, instead of checking `request` type.